concordium-consensus/src/Concordium/GlobalState/Persistent/TreeState.hs

```diff
@@ -490,37 +562,65 @@ instance (MonadLogger (PersistentTreeStateMonad bs m),
-                  (Nothing, Just _) -> logErrorAndThrowTS $ "Lost block that was stored as finalized" ++ show bh
-                  _ -> logErrorAndThrowTS $ "Lost block and finalization record" ++ show bh
+                case b of
+                  Nothing -> do
+                    deadBlocks <- use (blockTable . deadCache)
+                    return $! if memberDeadCache bh deadBlocks then Just TS.BlockDead else Nothing
+                  Just sb -> do
+                    fr <- readFinalizationRecord (sbFinalizationIndex sb)
+                    case fr of
+                      Just finr -> do
+                        block <- constructBlock sb
+                        return $ Just (TS.BlockFinalized block finr)
+                      Nothing -> logErrorAndThrowTS $ "Lost finalization record that was stored" ++ show bh
+
+    getRecentBlockStatus bh = do
+      st <- use (blockTable . liveMap . at' bh)
+      case st of
+        Just (BlockAlive bp) -> return $ TS.RecentBlock (TS.BlockAlive bp)
+        Just (BlockPending bp) -> return $ TS.RecentBlock (TS.BlockPending bp)
+        Nothing -> do
+            lf <- use lastFinalized
+            if bh == bpHash lf then do
+              lfr <- use lastFinalizationRecord
+              return $ TS.RecentBlock (TS.BlockFinalized lf lfr)
+            else do
+              b <- memberBlockTable bh
+              if b then
+                return TS.OlderThanLastFinalized
+              else do
+                  deadBlocks <- use (blockTable . deadCache)
+                  return $! if memberDeadCache bh deadBlocks then TS.RecentBlock TS.BlockDead else TS.Unknown
+
     makeLiveBlock block parent lastFin st arrTime energy = do
             blockP <- makePersistentBlockPointerFromPendingBlock block parent lastFin st arrTime energy
-            blockTable . at' (getHash block) ?=! BlockAlive blockP
+            blockTable . liveMap . at' (getHash block) ?=! BlockAlive blockP
             return blockP
-    markDead bh = blockTable . at' bh ?= BlockDead
+    markDead bh = do
+            blockTable . liveMap . at' bh .=! Nothing
+            blockTable . deadCache %=! insertDeadCache bh
     type MarkFin (PersistentTreeStateMonad bs m) = Maybe (StoredBlock (MPV m) (BlockStatePointer bs))
-    markFinalized bh fr = use (blockTable . at' bh) >>= \case
+    markFinalized bh fr = use (blockTable . liveMap . at' bh) >>= \case
             Just (BlockAlive bp) -> do
               st <- saveBlockState (_bpState bp)
-              blockTable . at' bh ?=! BlockFinalized (finalizationIndex fr)
+              -- NB: This only makes sense if no block lookups are done between
+              -- the call to this function and 'wrapUpFinalization'. This is
+              -- currently the case, and must remain the case in the future.
+              -- That is, finalization must remain atomic, or this handling, and
+              -- wrapUpFinalization must be changed.
```

Comment:
The documentation of `markFinalized` should be updated to reflect that the block is not properly finalized until `wrapupFinalization` is called, and thus looking up the block with `getBlockStatus` in the interim could give the wrong result.